// sqreen/test/lib/instrumentation/record.js
'use strict';
const Code = require('code');
const Lab = require('lab');
const lab = exports.lab = Lab.script();

const describe = lab.describe;
const it = lab.it;
const expect = Code.expect;

const Proxyquire = require('proxyquire');
const Utils = require('../../testUtils');
const RecordPath = '../../../lib/instrumentation/record';

describe('Record', () => {

    it('should create a record, fill it and close it', { plan: 3 }, (done) => {

        const eventStub = {
            writeEvent: function (type, content) {

                expect(type).to.equal('request_record');
                expect(content.req).to.not.exist();

                done();
            }
        };
        Utils.validateStub(require('../../../lib/events/index'), eventStub);

        const Record = Proxyquire(RecordPath, {
            '../events': eventStub
        });

        const req = {
            __sqreen_uuid: 'aa'
        };

        const record = new Record.Record(req, '127.0.0.1');

        expect(Record.STORE.get('aa')).to.equal(record);

        record.attack({
            rule_name: 'ruleName',
            test: true,
            block: false,
            infos: { a: 1 },
            time: new Date(),
            backtrace: (new Error('ruleName')).stack.split('\n')
        });

        const err = new Error('myErr');
        record.except({
            klass: Error.name,
            message: err.message,
            infos: {
                client_ip: null,
                args: []
            },
            rule_name: 'ruleName',
            time: new Date(),
            context: {
                backtrace: err.stack.split('\n')
            }
        });

        record.observe([['HTTP_CODE', 500, 1]]);

        record.close();
    });

    it('should create a record, fill it with only a metric and close it', { plan: 5 }, (done) => {

        const eventStub = {
            writeEvent: function (type, content) {

                done(new Error('should not have been called'));
            }
        };
        const metricStub = {
            addObservations: function (arg) {

                expect(arg).to.be.an.array();
                expect(arg).to.have.length(1);
                expect(arg[0]).to.be.an.array();
                expect(arg[0]).to.have.length(3);
                done();
            }
        };

        Utils.validateStub(require('../../../lib/events'), eventStub);
        Utils.validateStub(require('../../../lib/metric'), metricStub);

        const Record = Proxyquire(RecordPath, {
            '../events': eventStub,
            '../metric': metricStub
        });

        const req = {
            __sqreen_uuid: 'aa'
        };

        const record = new Record.Record(req, '127.0.0.1');

        expect(Record.STORE.get('aa')).to.equal(record);

        record.observe([['HTTP_CODE', 500, 1]], new Date());

        record.close();
    });
});
