// sqreen/lib/instrumentation/record.js
/**
 * Copyright (c) 2016 Sqreen. All Rights Reserved.
 * Please refer to our terms for more information: https://www.sqreen.io/terms.html
 */
'use strict';
const ReportUtil = require('./reportUtil');
const Util = require('../util');
const Rules = require('../rules');
const Metric = require('../metric');
const Events = require('../events');
const TYPE = require('../enums/events').TYPE;
const FORMAT_VERSION = '20171208';

const STORE = module.exports.STORE = new Map();

const Record = class {

    constructor(req, client_ip) {

        this.version = FORMAT_VERSION;
        this.rulespack_id = Rules.rulespack;
        this.client_ip = client_ip || Util.getXFFOrRemoteAddress(req);
        this.request = {};
        // this.response = {}; // TODO: future
        this.observed = {
            attacks: [],
            sqreen_exceptions: [],
            observations: []
        };
        this.sdk  = [];

        this.req = req;
        STORE.set(this.req.__sqreen_uuid, this);
    }

    attack(atk) {

        this.observed.attacks.push(atk);
    }

    except(exc) {

        this.observed.sqreen_exceptions.push(exc);
    }

    observe(observationList, date) {

        date = date || new Date();
        for (let i = 0; i < observationList.length; ++i) {
            this.observed.observations.push({
                category: observationList[i][0],
                key: observationList[i][1],
                value: observationList[i][2],
                time: date
            });
        }
    }

    addSDK(name, args) {

        const time = new Date();
        this.sdk.push({ time, name, args });
    }

    shouldReport() {

        return this.observed.attacks.length > 0 || this.observed.sqreen_exceptions.length > 0;
    }

    reportMetric() {

        this.observed.observations.forEach((x) => {

            Metric.addObservations([[x.category, x.key, x.value]], x.time);
        });
    }

    report() {

        if (this.shouldReport()) {
            this.request = ReportUtil.mapRequest(this.req);
            this.req = undefined;
            Events.writeEvent(TYPE.REQUEST_RECORD, this);
            return;
        }
        return this.reportMetric();
    }

    close() {

        const uuid = this.req.__sqreen_uuid;
        this.report();
        STORE.delete(uuid);
    }
};

module.exports.Record = Record;
module.exports.lazyGet = function (req) {

    req = req || {};
    if (req.__sqreen_uuid === undefined) {
        return null;
    }
    const current = STORE.get(req.__sqreen_uuid);
    if (current === undefined) {
        return new Record(req);
    }
    return current;
};
